python/cudf_polars/cudf_polars/containers/column.py

```diff
@@ -129,11 +140,46 @@ def astype(self, dtype: plc.DataType) -> Column:
         This only produces a copy if the requested dtype doesn't match
         the current one.
         """
-        if self.obj.type() != dtype:
-            return Column(plc.unary.cast(self.obj, dtype), name=self.name).sorted_like(
-                self
-            )
-        return self
+        if self.obj.type() == dtype:
+            return self
+
+        if dtype.id() == plc.TypeId.STRING or self.obj.type().id() == plc.TypeId.STRING:
+            return Column(self._handle_string_cast(dtype))
+        else:
+            result = Column(plc.unary.cast(self.obj, dtype))
+            if is_order_preserving_cast(self.obj.type(), dtype):
+                return result.sorted_like(self)
+            return result
+
+    def _handle_string_cast(self, dtype: plc.DataType) -> plc.Column:
+        if dtype.id() == plc.TypeId.STRING:
+            if is_floating_point(self.obj.type()):
+                return from_floats(self.obj)
+            else:
+                return from_integers(self.obj)
+        else:
+            if is_floating_point(dtype):
+                floats = is_float(self.obj)
+                if not plc.interop.to_arrow(
+                    plc.reduce.reduce(
+                        floats,
+                        plc.aggregation.all(),
+                        plc.DataType(plc.TypeId.BOOL8),
+                    )
+                ).as_py():
+                    raise InvalidOperationError("Conversion from `str` failed.")
+                return to_floats(self.obj, dtype)
+            else:
+                integers = is_integer(self.obj)
+                if not plc.interop.to_arrow(
+                    plc.reduce.reduce(
+                        integers,
+                        plc.aggregation.all(),
+                        plc.DataType(plc.TypeId.BOOL8),
+                    )
+                ).as_py():
+                    raise InvalidOperationError("Conversion from `str` failed.")
+                return to_integers(self.obj, dtype)
```

Comment:
How does this match up with polars behaviour for casting?

For example, in polars, there are, for numerical casts, strict/non-strict (out of range raises or is nullified), and also wrapping/non-wrapping (takes precedence over strictness) "wrapping" casts match up with what libcudf does, I think.

For string casting, wrapping makes no sense, so strict means "raise" and non-strict means "replace with null". What does libcudf do?

You can see what behaviour the cast node requested (in `translate.py`) by looking at its `options`. `0 => strict`, `1 => non-strict`, `2 => wrapping`.

I am happy to defer this to a followup, since this code is an improvement. But I would like to at least document how this differs from polars CPU.

Reply:
The current code implements `strict=True`, which is the default. `strict=False` would result in null values. The libcudf code does not automatically replace strings that won't convert with nulls, instead the behavior appears to be undefined. We can:

1. Block `strict=False` for now.
2. Implement `strict=False` by scattering nulls into the result based on the `is_integer` mask.
3. Something else I haven't thought of?

Reply:
4. Open an issue to implement the other strictness cases, and leave this as is for now.

Reply:
https://github.com/rapidsai/cudf/issues/17244